admin/acceptance.py

```diff
@@ -528,19 +577,196 @@ def start_cluster(self, reactor):
+                    # Destroy a node if we failed to provision it.
+                    self._destroy_node(node)
+                    return failure
+
+                d.addErrback(error)
+                # Make sure too return the node.
+                d.addCallback(lambda _: node)
+                return d
+
+            d.addCallback(provision)
+
+            # Log and discard a failure to keep looping.
+            d.addErrback(write_failure)
+            return d
+
+        d = loop_until(reactor, create_attempt, repeat(0, retries))
+
+        def error(failure):
+            print "Failed to provision node {}".format(name)
+            return failure
+
+        d.addErrback(error)
+        return d
+
+    def _create_nodes(self, reactor, names):
+        """
+        Create and provision a number of nodes with the given names.
+
+        :param reactor: The reactor.
+        :param names: The names.
+        :type names: list of str
+        :return: List of Deferreds each firing when the corresponding node
+            is created and provisioned.
+        """
+        reactor.suggestThreadPoolSize(len(names))
+        return [
+            self._create_and_provision(reactor, name) for name in names
+        ]
+
+    def _make_node_name(self, tag, index):
+        """
+        Construct a node name out of various components.
+        """
+        return "%s-%s-%s-%d" % (
+            self.identity.prefix, self.creator, tag, index,
+        )
+
+    def _add_node_to_cluster(self, reactor, cluster, node, index):
+        """
+        Configure the given node as an agent node for the given cluster.
+
+        :param reactor: The reactor.
+        :param reactor: The cluster.
+        :param Node node: The node.
+        :param int index: The index of the ndoe in the cluster.
+        :return: Deferred that fires with the :param:`node` when it is
+            configured.
+        """
+        node_certnkey = cluster.certificates.add_node(index)
+        commands = configure_node(
+            cluster,
+            node,
+            node_certnkey,
+            self.dataset_backend_configuration,
+            'libcloud',
+            logging_config=self.config.get('logging'),
+        )
+        d = perform(make_dispatcher(reactor), commands)
+
+        def add_node(ignored):
+            if node is not cluster.control_node:
+                cluster.all_nodes.append(node)
+            cluster.agent_nodes.append(node)
+            return node
+
+        def error(failure):
+            print "Failed to configure node {}".format(node.name)
+            write_failure(failure)
+            if node is not cluster.control_node:
+                self._destroy_node(node)
+            # Failure to add a node to a cluster is not fatal.
+            # As long as we have the control node we can claim
+            # the partial success.
+            return failure
+
+        d.addErrback(error)
+        d.addCallback(add_node)
```

Comment:
Although I am not answering the question - I will leave that to Andriy - note that using `addCallbacks` is semantically different.  As the code is now, if the `error` function returns any object other than a `Failure`, it will then call `add_node`.

Changing to `addCallbacks` would make it call one or the other, never both.

Given that the `error` function appears to always return a `Failure`, they are probably effectively the same.  So, yes, it should be changed unless `error` could potentially offer a path to redemption (from error state back to normal flow).